src/modules/rlm_sql/drivers/rlm_sql_postgresql/rlm_sql_postgresql.c

```diff
@@ -296,14 +296,46 @@ static CC_HINT(nonnull) sql_rcode_t sql_query(rlm_sql_handle_t *handle, UNUSED r
 					      char const *query)
 {
 	rlm_sql_postgres_conn_t *conn = handle->conn;
+	struct timeval timeout = {config->query_timeout, 0};
+	int sockfd;
+	fd_set read_fd;
 	ExecStatusType status;
 	int numfields = 0;
+	PGresult *tmp_result;
 
 	if (!conn->db) {
 		ERROR("rlm_sql_postgresql: Socket not connected");
 		return RLM_SQL_RECONNECT;
 	}
 
+	sockfd = PQsocket(conn->db);
+	if (sockfd < 0) {
+		ERROR("rlm_sql_postgresql: Unable to obtain socket: %s", PQerrorMessage(conn->db));
+		return RLM_SQL_RECONNECT;
+	}
+
+	if (!PQsendQuery(conn->db, query)) {
+		ERROR("rlm_sql_postgresql: Failed to send query: %s", PQerrorMessage(conn->db));
+		return RLM_SQL_RECONNECT;
+	}
+
+	/*
+	 * We try to avoid blocking by waiting until the driver indicates that
+         * the result is ready or our timeout expires
+	 */
+	while (PQisBusy(conn->db)) {
+		FD_ZERO(&read_fd);
+		FD_SET(sockfd, &read_fd);
+		if (!select(sockfd + 1, &read_fd, NULL, NULL, config->query_timeout ? &timeout : NULL)) {
```

Comment:
Should probably check for -1 return code here too, mostly for EAGAIN and EINTR.  Not 100% sure how v3.0.x handles signals.  No point in making the change for v4.0.x, but just to be safe for v3.0.x

Reply:
What is the expected behaviour here for EINTR amd EAGAIN? At the moment we would stay in the loop and retry the select (unless a handler is defined) which seems pretty standard.

It might be a bit drastic to fail the request on receipt of a signal. EAGAIN is interesting as it's sometimes used as a proxy for ENOMEM, so is it best to repeat or give up...

Linux man 2 select has the following which suggests repeating in both cases, but is that best here:

> On some other UNIX systems, select() can fail with the error EAGAIN if the system fails to allocate kernel-internal resources, rather than ENOMEM as Linux does.  POSIX specifies this error for poll(2), but not for select().  Portable programs may wish to check for EAGAIN and loop, just as with EINTR.

Reply:
EINTR should definitely be repeated.  EAGAIN/ENOMEM require previously used resources being released in order to make progress otherwise we'd essentially busy loop, so we probably want to bail out of the loop in that case.

Regarding continuing in the loop body, I'm just a bit paranoid about calling PQconsumeInput where no data is available.